src/main/java/build/buf/protovalidate/internal/constraints/ConstraintCache.java

```diff
@@ -127,18 +155,101 @@ public List<CompiledProgram> compile(
             continue;
           }
         }
-        Ast residual = finalEnv.residualAst(astExpression.ast, evalResult.getEvalDetails());
+        Ast residual = ruleEnv.residualAst(rule.astExpression.ast, evalResult.getEvalDetails());
         programs.add(
-            new CompiledProgram(finalEnv.program(residual, rulesOption), astExpression.source));
+            new CompiledProgram(ruleEnv.program(residual, globals), rule.astExpression.source));
       } catch (Exception e) {
         programs.add(
             new CompiledProgram(
-                finalEnv.program(astExpression.ast, rulesOption), astExpression.source));
+                ruleEnv.program(rule.astExpression.ast, globals), rule.astExpression.source));
       }
     }
     return Collections.unmodifiableList(programs);
   }
 
+  private @Nullable List<CelRule> compileRule(
+      FieldDescriptor fieldDescriptor,
+      boolean forItems,
+      FieldDescriptor constraintFieldDesc,
+      Message message)
+      throws CompilationException {
+    if (descriptorMap.containsKey(constraintFieldDesc)) {
+      return descriptorMap.get(constraintFieldDesc);
+    }
```

Comment:
Oops, good point. (I didn't actually change this line, it just got relocated, but there's no sense in keeping it like this, since it's not relying on `null` values in the map or anything like that.)